### PR TITLE
Keypoint occlusion

### PR DESCRIPTION
**Pull request by StevenBorkman**
# Peer Review Information:
Fixes for both 1160 & 1161. Labeled assets with all of their keypoints completely off screen are not reported to the captures.json file. If a labeled asset is at least partially on screen (or more precise at least 1 of its keypoints are on screen) then all the keypoints will be marked with the following values: 0 => keypoint does not exist in the model or is off screen, 1 => keypoint exists and is on screen, but is occluded by some other object, 2 => keypoint exists and is visible.

New correctness tests have been included to test all of these states.

## Editor / Package versioning:
**Editor Version Target (i.e. 19.3, 20.1)**: 2019.3

## Dev Testing:
**Tests Added**: 

**Core Scenario Tested**: 

**At Risk Areas**: 

**Notes + Expectations**: 

## Checklist
- [x] - Updated docs
- [x] - Updated changelog

### CODE REVIEW

File: com.unity.perception/Runtime/GroundTruth/Labelers/KeypointLabeler.cs [hunk 1]
```diff
@@ -86,33 +92,95 @@ protected override void Setup()
...
+            asyncAnnotation.ReportValues(m_ToReport);
         }
 
         /// <inheritdoc/>
         protected override void OnBeginRendering()
         {
-            var reporter = perceptionCamera.SensorHandle.ReportAnnotationAsync(m_AnnotationDefinition);
+            m_CurrentFrame = Time.frameCount;
+
+            m_Keypoints[m_CurrentFrame] = new Dictionary<uint, KeypointEntry>();
```
Comment (JonathanHUnity) [inline com.unity.perception/Runtime/GroundTruth/Labelers/KeypointLabeler.cs:171]: For performance reasons we should avoid allocating managed objects on a per-frame basis. It's a death by a thousand cuts kind of issue. Give some thought to how you could use the Native* types by allocating all the memory you need up front. If you don't see a path, we can leave this exercise for another time.

**Review by JonathanHUnity — CHANGES_REQUESTED** [review]
Great. Couple notes left to look at

File: com.unity.perception/CHANGELOG.md [hunk 2]
```diff
@@ -19,6 +19,14 @@ Scenario serialization has been updated to include scalar values on randomizers
 
 Added new ScenarioBase virtual lifecycle hooks: OnAwake, OnStart, OnComplete, and OnIdle.
 
+Keypoint occlusion has been added, no keypoint information will be recorded for a labeled asset completely out of the camera's frustum. 
+If a labeled asset is at least partially visible all keypoints will be annotated with the following state values:<br> 
+0: Off of screen or does not exist in model<br>
+1: Exists in model and in camera's frustum but occluded by another object<br>
+2: On screen and visible.
+
+New keypoint tests have been added to test keypoint state
```
[the remark this answers is not in the document]
Comment (StevenBorkman) [reply com.unity.perception/CHANGELOG.md]: Makes sense. Updated.

File: com.unity.perception/Runtime/GroundTruth/Labelers/KeypointLabeler.cs [hunk 3]
```diff
@@ -86,33 +92,95 @@ protected override void Setup()
...
+
+        void OnInstanceSegmentationImageReadback(int frameCount, NativeArray<Color32> data, RenderTexture renderTexture)
+        {
+            if (!m_AsyncAnnotations.TryGetValue(frameCount, out var asyncAnnotation))
+                return;
+
+            m_AsyncAnnotations.Remove(frameCount);
+
+            if (!m_Keypoints.TryGetValue(frameCount, out var keypoints))
+                return;
```
[the remark this answers is not in the document]
Comment (StevenBorkman) [reply com.unity.perception/Runtime/GroundTruth/Labelers/KeypointLabeler.cs]: Ok, implemented suggestions.

File: com.unity.perception/Runtime/GroundTruth/Labelers/KeypointLabeler.cs (hunk 1 above)
Comment (StevenBorkman) [reply com.unity.perception/Runtime/GroundTruth/Labelers/KeypointLabeler.cs]: I looked around at this. I couldn't really find a way to do this without going fully into ECS and turning the class into a component. So I think we may have to punt on this for right now and come back to it.

**Review by JonathanHUnity — APPROVED** [review]
Nice

**Pull request merged**